### ISSUE TITLE
Overloaded functions with implementation don't specialize

### ISSUE THREAD

**Issue by fluggo** [microsoft/pyright#301]
**Describe the bug**
When providing the implementation of an overloaded function, such as in the documentation for the `@overload` attribute, Pyright uses the types from the function implementation instead of the appropriate overload specialization.

**To Reproduce**
Here, Pyright flags the function call because None is not assignable to datetime. It does this because it uses the type definitions from the implementation of the function instead of its overloads.

```python
from typing import overload, Optional
from datetime import datetime, timezone, timedelta

@overload
def from_json_timestamp(ts: int) -> datetime:
    ...

@overload
def from_json_timestamp(ts: None) -> None:
    ...

def from_json_timestamp(ts: Optional[int]) -> Optional[datetime]:
    '''
    Returns a native timezone-aware datetime from the given POSIX millisecond timestamp.
    '''
    return None if ts is None else (datetime(1970, 1, 1, tzinfo=timezone.utc) + timedelta(milliseconds=ts))

result: datetime = from_json_timestamp(2418049)
```

**Expected behavior**
Since Pyright can see that the input to the function is an int, it should have chosen the specialization that only returns datetime.

**VS Code extension or command-line**
This is the VS Code extension version 1.0.80.

**Comment by fluggo** [comment]
This version seems to work, but it means that the types in the implementation aren't checked:

```python
from typing import overload, Optional
from datetime import datetime, timezone, timedelta

@overload
def from_json_timestamp(ts: int) -> datetime:
    ...

@overload
def from_json_timestamp(ts: None) -> None:
    ...

def from_json_timestamp(ts: ...) -> ...:
    '''
    Returns a native timezone-aware datetime from the given POSIX millisecond timestamp.
    '''
    return None if ts is None else (datetime(1970, 1, 1, tzinfo=timezone.utc) + timedelta(milliseconds=ts))

result = from_json_timestamp(2418049)
print(result.hour)
```

**Comment by erictraut** [comment]
Thanks for the bug report. This was due to a regression I recently introduced. I've fixed the problem and plugged the hole in the unit tests so a similar regression won't go uncaught in the future. A fix will be in the next release.

**Comment by fluggo** [comment]
Thanks much, Eric!

**Comment by erictraut** [comment]
This is now fixed in 1.0.81, which I just published
